### PR TITLE
Add plexon2 support

### PR DESCRIPTION
**Pull request by JuliaSprenger**
This is a draft to add support for the new Plexon pl2 support to neo as among other sources requested in https://github.com/NeuralEnsemble/python-neo/issues/1124

The RawIO will be based on the PyPl2 scripts and corresponding dlls provided by Plexon (see https://gitlab.com/penalab/pypl2 and https://github.com/juliasprenger/pypl2)

This PR requires the test file to be available on GIN, so https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/pulls/97/files needs to be merged first.

### CODE REVIEW

**Comment by JuliaSprenger** [comment]
This PR is in principle ready for review. Test files need to be merged for tests to pass.

I didn't have any test files with multiple analog signal fragments present, if anybody has a file with this feature it would be great to be tested before merging.

@brian-lau @KonKob @vincisLab

**Comment by JuliaSprenger** [comment]
TODOs:
- [ ] Move the dll to an external repo and fetch it.

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+        >>> spike_timestamp = r.get_spike_timestamps(spike_channel_index=0, t_start=None,
+                            t_stop=None)
+        >>> spike_times = r.rescale_spike_timestamp(spike_timestamp, 'float64')
+        >>> ev_timestamps, _, ev_labels = r.get_event_timestamps(event_channel_index=0)
+
+    """
+    extensions = ['pl2']
+    rawmode = 'one-file'
+
+    def __init__(self, filename='', pl2_dll_file_path=None):
```
Comment (bendichter) [inline neo/rawio/plexon2rawio/plexon2rawio.py:56]: why have the default value of filename be an empty string? Shouldn't this be a required arg?

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+        self.filename = pathlib.Path(filename)
+
+        if (not self.filename.exists()) or (not self.filename.is_file()):
+            raise ValueError(f'{self.filename} is not a file.')
+
+        BaseRawIO.__init__(self)
+
+        # download default PL2 dll once if not yet available
+        if pl2_dll_file_path is None:
+            pl2_dll_folder = pathlib.Path .home() / '.plexon_dlls_for_neo'
```
Comment (bendichter) [inline neo/rawio/plexon2rawio/plexon2rawio.py:74]: Maybe this should live in the package directory. That way when you uninstall the package you uninstall the dll

Comment (JuliaSprenger) [reply neo/rawio/plexon2rawio/plexon2rawio.py:74]: This approach might cause problems if the installation was performed with different right than the execution. The package directory might be read-only. Creating a folder for the dll in the home directory should not fail due to permissions and is the same location used also for cloning of the `ephy_testing_files` repository. I agree that there should be a mechanism to register these additional folders in the pyproject.toml to be also removed upon deinstallation of neo, but didn't find the right option yet. Let me know if you have seen something like this already.

**Comment by JuliaSprenger** [comment]
Tests are not completing at the moment.
- [ ] fix Plexon2IO tests

**Comment by JuliaSprenger** [comment]
@samuelgarcia The failing tests are not related to the new IO. Could you have a look at the proposed IO?

**Comment by JuliaSprenger** [comment]
Hi @samuelgarcia Ready for review!

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+            pl2_dll_folder.mkdir(exist_ok=True)
+            pl2_dll_file_path = pl2_dll_folder / 'PL2FileReader.dll'
+
+            if pl2_dll_file_path.exists():
+                warnings.warn(f'Using cached plexon dll at {pl2_dll_file_path}')
+            else:
+                dist = urlopen('https://github.com/Neuralensemble/pypl2/blob/master/bin/PL2FileReader.dll?raw=true')
```
Comment (samuelgarcia) [inline neo/rawio/plexon2rawio/plexon2rawio.py:82]: I think we can also use `https://raw.githubusercontent.com/...` no ?

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+                    achannel_info = self.pl2reader.pl2_get_analog_channel_info_by_name(achannel_name)
+                    for key in signal_array_annotation_keys:
+                        signal_array_annotations[key].append(getattr(achannel_info, key))
+
+                seg_ann['signals'][stream_idx]['__array_annotations__'] = signal_array_annotations
+
+
+    def _segment_t_start(self, block_index, seg_index):
+        # this must return an float scale in second
+        # this t_start will be shared by all object in the segment except AnalogSignal
```
Comment (samuelgarcia) [inline neo/rawio/plexon2rawio/plexon2rawio.py:279]: This comments is from the example no ?

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+        for i, channel_idx in enumerate(channel_indexes):
+            channel_name = stream_channels['name'][channel_idx]
+
+            # use previously loaded channel data if possible
+            if channel_name in self._analogsignal_cache:
+                values = self._analogsignal_cache[channel_name]
+            else:
+                res = self.pl2reader.pl2_get_analog_channel_data_by_name(channel_name)
+                fragment_timestamps, fragment_counts, values = res
+                self._analogsignal_cache[channel_name] = values
```
Comment (samuelgarcia) [inline neo/rawio/plexon2rawio/plexon2rawio.py:353]: I guess `pl2_get_analog_channel_data_by_name` is very slow...

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+            timestamp_frequency = self.pl2reader.pl2_file_info.m_TimestampFrequency
+            lim0 = int(t_start * timestamp_frequency)
+            lim1 = int(t_stop * self.pl2reader.pl2_file_info.m_TimestampFrequency)
+            time_mask = (spike_timestamps >= lim0) & (spike_timestamps <= lim1)
+
+            # limits are with respect to segment t_start and not to time 0
+            lim0 -= self.pl2reader.pl2_file_info.m_StartRecordingTime
+            lim1 -= self.pl2reader.pl2_file_info.m_StartRecordingTime
+        else:
+            time_mask = slice(None, None)
```
Comment (samuelgarcia) [inline neo/rawio/plexon2rawio/plexon2rawio.py:433]: To be safe here if we correct a bug one day I would share this piece of code with the one in _get_spike_timestamps

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+                wf_gain = schannel_info.m_CoeffToConvertToUnits
+                wf_offset = 0.  # A waveform offset is not provided in PL2 files
+                wf_left_sweep = schannel_info.m_PreThresholdSamples
+                wf_sampling_rate = schannel_info.m_SamplesPerSecond
+                spike_channels.append((unit_name, unit_id, wf_units, wf_gain,
+                                      wf_offset, wf_left_sweep, wf_sampling_rate))
+
+            # pre-loading spiking data
+            schannel_name = schannel_info.m_Name.decode()
+            self._spike_channel_cache[schannel_name] = self.pl2reader.pl2_get_spike_channel_data_by_name(schannel_name)
```
Comment (samuelgarcia) [inline neo/rawio/plexon2rawio/plexon2rawio.py:166]: I think I would do this on demand and not in init.

**Comment by samuelgarcia** [comment]
Hi julia.
Thank you for this incredibly tedious task of wrapping a windows ddl here.
This is heroic.
My guess is that implementing from scratch the reader with good specification would be simpler :)

This is really hard to really give feedback of the ctype wrapper layer because I do not known the ddl.
I made some generic feedback a some places you can fix or not.

Maybe you could do some clean on the comments hinerited from exmaplerawio to keep only relevent ones.

File: neo/rawio/plexon2rawio/plexon2rawio.py (hunk 3 above)
Comment (JuliaSprenger) [reply neo/rawio/plexon2rawio/plexon2rawio.py:82]: Yes, you are right. This way we can remove the `/blob` and `?raw=true`
```suggestion
                dist = urlopen('https://raw.githubusercontent.com/Neuralensemble/pypl2/master/bin/PL2FileReader.dll')
```

File: neo/rawio/plexon2rawio/plexon2rawio.py (hunk 7 above)
Comment (JuliaSprenger) [reply neo/rawio/plexon2rawio/plexon2rawio.py]: Good idea, I moved the loading into cache in the respective methods.

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+            echannel_info = self.pl2reader.pl2_get_digital_channel_info(i)
+
+            # only consider active channels
+            if not echannel_info.m_ChannelEnabled:
+                continue
+
+            # event channels are characterized by (name, id, type), with type in ['event', 'epoch']
+            channel_name = echannel_info.m_Name.decode()
+            event_channels.append((channel_name, echannel_info.m_Channel, 'event'))
+            self._event_channel_cache[channel_name] = self.pl2reader.pl2_get_digital_channel_data_by_name(channel_name)
```
[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply neo/rawio/plexon2rawio/plexon2rawio.py]: Done.

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+
+        # convert ctypes datetime objects to datetime.datetime objects for annotations
+        from .pypl2.pypl2lib import tm
+        for anno_key, anno_value in block_info.items():
+            if isinstance(anno_value, tm):
+                tmo = anno_value
+                # invalid datetime information if year is <1
+                if tmo.tm_year != 0:
+                    dt = datetime(year=tmo.tm_year, month=tmo.tm_mon, day=tmo.tm_mday, hour=tmo.tm_hour,
+                                  minute=tmo.tm_min, second=tmo.tm_sec, microsecond=block_info['m_CreatorDateTimeMilliseconds'])
```
[the remark this answers is not in the document]
Comment (JuliaSprenger) [reply neo/rawio/plexon2rawio/plexon2rawio.py]: Good catch, I fixed it by adding a conversion factor.

File: neo/rawio/plexon2rawio/plexon2rawio.py (hunk 4 above)
Comment (JuliaSprenger) [reply neo/rawio/plexon2rawio/plexon2rawio.py]: Yes, i kept some of the comments from the example io, but this one is not applicable any more. I removed it.

File: neo/rawio/plexon2rawio/plexon2rawio.py (hunk 5 above)
Comment (JuliaSprenger) [reply neo/rawio/plexon2rawio/plexon2rawio.py:347]: I did some test runs of this on a multi-GB pl2 file and it seems that the `_by_name` method is even a tiny bit faster than accessing the channel by its id. I don't know what exactly is happening in the back as the dll is closed source, but I would stick to the `pl2_get_analog_channel_dat_by_name` method here for now.

**Comment by samuelgarcia** [comment]
Hi Julia.
Tell me when ready for merge.

**Comment by JuliaSprenger** [comment]
Hi @samuelgarcia Thanks for the feedback. I addressed your comments and in principle only one is still open regarding loading of data by the channel name. I cleaned some of the comments that remained from the example io, but still many of them might be useful later on to understand why certain methods have certain return values and signatures. Can you check again and see if you are happy with the current changes? From my side this is ready for merging.

**Comment by samuelgarcia** [comment]
OK lets go go for merging.
@bendichter @Chetank99  and @JuliaSprenger : total apologies for having so much delays.

File: neo/rawio/plexon2rawio/plexon2rawio.py [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+        if (not self.filename.exists()) or (not self.filename.is_file()):
+            raise ValueError(f'{self.filename} is not a file.')
+
+        BaseRawIO.__init__(self)
+
+        # download default PL2 dll once if not yet available
+        if pl2_dll_file_path is None:
+            pl2_dll_folder = pathlib.Path .home() / '.plexon_dlls_for_neo'
+            pl2_dll_folder.mkdir(exist_ok=True)
+            pl2_dll_file_path = pl2_dll_folder / 'PL2FileReader.dll'
```
Comment (Chetank99) [inline neo/rawio/plexon2rawio/plexon2rawio.py:79]: Hi @JuliaSprenger 
This pull request was super helpful!!
Just one thing, while executing Plexon2RawIO(). I got the error:
OSError: [WinError 193] %1 is not a valid Win32 application
and
OSError: Error: Can't load PL2FileReader.dll at: C:\Users\admin.plexon_dlls_for_neo\PL2FileReader.dll. PL2FileReader.dll is bundled with the C++ PL2 Offline Files SDK located on the Plexon Inc website: www.plexon.com Contact Plexon Support for more information: support@plexon.com

I then realized my pypl2 SDK folder had another dll file called "PL2FileReader64.dll"
so after replace here "PL2FileReader.dll" with "PL2FileReader64.dll", it started running smoothly.

Just wanted to keep a note here if anyone else encounters those errors and wants to move ahead

**Pull request merged**